js/id/actions/merge_remote_changes.js

```diff
@@ -0,0 +1,152 @@
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({loc: remote.loc});
+        }
+
+        conflicts.push(t('merge_remote_changes.conflict.location'));
+        return;  // fail merge
+    }
+
+    function mergeRemoteChildren(target) {
+        if (!target) return;
+
+        if (_.isEqual(target.nodes, remote.nodes)) {
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({nodes: remote.nodes});
+        }
+
+        var o = base.nodes || [],
+            a = local.nodes || [],
+            b = remote.nodes || [],
+            nodes = [],
+            hunks = Diff3.diff3_merge(a, o, b, true);
+
+        for (var i = 0, imax = hunks.length; i !== imax; i++) {
+            var hunk = hunks[i];
+            if (hunk.ok) {
+                nodes.push.apply(nodes, hunk.ok);
+            }
+            else {
+                // for all conflicts, we can assume c.a !== c.b
+                // because `diff3_merge` called with `true` option to exclude false conflicts..
+                var c = hunk.conflict;
+                if (_.isEqual(c.o, c.a)) {  // only changed remotely
+                    nodes.push.apply(nodes, c.b);
+                }
+                else if (_.isEqual(c.o, c.b)) {  // only changed locally
+                    nodes.push.apply(nodes, c.a);
+                }
+                else {       // changed both locally and remotely
+                    conflicts.push(t('merge_remote_changes.conflict.nodelist'));
+                    return;  // fail merge..
+                }
+            }
+        }
+
+        return target.update({nodes: nodes});
+    }
+
+    function mergeRemoteMembers(target) {
+        if (!target) return;
+
+        if (_.isEqual(target.members, remote.members)) {
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({members: remote.members});
+        }
+
+        conflicts.push(t('merge_remote_changes.conflict.memberlist'));
+        return;  // fail merge
+    }
+
+    function mergeRemoteTags(target) {
+        if (!target) return;
+
+        if (_.isEqual(target.tags, remote.tags)) {
+            return target;
+        }
+        if (option === 'force_remote') {
+            return target.update({tags: remote.tags});
+        }
+
+        var keys = _.reject(_.union(_.keys(base.tags), _.keys(remote.tags)), ignoreKey),
+            tags = _.clone(target.tags),
+            changed = false,
+            fail = false;
+
+        function ignoreKey(k) {
+            return k.indexOf('tiger:') === 0 || _.contains(iD.data.discarded, k);
+        }
+
+        for (var i = 0, imax = keys.length; i !== imax; i++) {
+            var k = keys[i];
+            if (remote.tags[k] !== base.tags[k]) {  // tag modified remotely..
+                if (target.tags[k] && target.tags[k] !== remote.tags[k]) {
```

Comment:
I'm probably going to change this to raise the conflict rather than silently restoring the tag, since we are probably moving in the direction of giving the user buttons to resolve the conflict anyway.

I think the most common scenario here goes something like:   UserA sees an address that he knows is wrong and deletes it, meanwhile UserB knows the correct street address and enters it. I'm ok with telling UserA about the conflict.